browser/src/Renderer/WebGL/WebGLAtlas.ts

```diff
@@ -0,0 +1,141 @@
+const TEXTURE_SIZE = 512
+
+export interface IWebGLAtlasOptions {
+    fontFamily: string
+    fontSize: string
+    lineHeight: number
+    devicePixelRatio: number
+    subpixelDivisor: number
+}
+
+export interface WebGLGlyph {
+    width: number
+    height: number
+    textureWidth: number
+    textureHeight: number
+    textureU: number
+    textureV: number
+    variantOffset: number
+    subpixelWidth: number
+}
+
+export class Atlas {
+    private glyphCanvas: HTMLCanvasElement
+    private glyphCtx: CanvasRenderingContext2D
+    private glyphPadding = 2
+    private glyphs = new Map<string, Map<number, WebGLGlyph>>()
+    private nextX = 0
+    private nextY = 0
+    private textureChangedSinceLastUpload = false
+    private texture: WebGLTexture
+    private textureSize: number
+    private uvScale: number
+
+    constructor(private gl: WebGL2RenderingContext, private options: IWebGLAtlasOptions) {
+        this.textureSize = TEXTURE_SIZE * options.devicePixelRatio
+        this.uvScale = 1 / this.textureSize
+
+        this.glyphCanvas = document.createElement("canvas")
+        this.glyphCanvas.style.backgroundColor = "red"
+        this.glyphCanvas.width = this.textureSize
+        this.glyphCanvas.height = this.textureSize
+        this.glyphCtx = this.glyphCanvas.getContext("2d", { alpha: false })
+        this.glyphCtx.font = `${this.options.fontSize} ${this.options.fontFamily}`
+        this.glyphCtx.fillStyle = "white"
+        // this.glyphCtx.textBaseline = "bottom"
+        this.glyphCtx.textBaseline = "top"
+        this.glyphCtx.scale(options.devicePixelRatio, options.devicePixelRatio)
+        this.glyphCtx.imageSmoothingEnabled = false
+
+        this.texture = gl.createTexture()
+        gl.bindTexture(gl.TEXTURE_2D, this.texture)
+        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.LINEAR)
+        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_S, gl.CLAMP_TO_EDGE)
+        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_T, gl.CLAMP_TO_EDGE)
+        gl.texImage2D(
+            gl.TEXTURE_2D,
+            0,
+            gl.RGBA,
+            this.textureSize,
+            this.textureSize,
+            0,
+            gl.RGBA,
+            gl.UNSIGNED_BYTE,
+            this.glyphCanvas,
```

Comment:
This is cool - if I understand correctly, this is how we can take our 'Canvas' that we render characters / glyphs to, and supply it as a texture that we can read from the shaders?

Reply:
That's exactly what we're doing here :) This way we get the nice font rendering from Chrome while at the same time being able to run our editor through WebGL.

The only trade-off is that we need to manually manage our subpixel offsets: We need to manually manage the kerning.

Without additional effort, there is only one version of a character in the atlas, typically aligned to the full pixel boundary. With non-integer character widths however, this would probably ruin the kerning in many cases. A character might at some point need to start at 0.7 pixels offset within the pixel grid. To approximate the correct position, we have different variants of each rendered character in the atlas, each offset by a different fraction of a full pixel. Then we use the one that is closest to what we need at the given position in the editor panel. The number of different variants is controlled by `subpixelDivisor` in this file.

This idea was completely stolen from xray, BTW! All credit goes to @nathansobo :)